src/auth.js

```diff
@@ -0,0 +1,159 @@
+import xdr from './xdr';
+
+import { StrKey } from './strkey';
+import { Keypair } from './keypair';
+import { hash } from './hashing';
+
+import { Address } from './address';
+import { nativeToScVal } from './scval';
+
+/**
+ * This builds an authorization entry that indicates to
+ * {@link Operation.invokeHostFunction} that a particular identity (i.e. signing
+ * {@link Keypair}) approves the execution of an invocation tree (i.e. a
+ * simulation-acquired {@link xdr.SorobanAuthorizedInvocation}) on a particular
+ * network (uniquely identified by its passphrase, see {@link Networks}) until a
+ * particular ledger sequence is reached.
+ *
+ * This enables building an {@link xdr.SorobanAuthorizationEntry} without
+ * worrying about how to combine {@link buildAuthEnvelope} and
+ * {@link buildAuthEntry}, while those allow advanced, asynchronous, two-step
+ * building+signing of the authorization entries.
+ *
+ * This one lets you pass a either a {@link Keypair} or a callback function to
+ * handle signing the envelope hash.
+ *
+ * @param {Keypair} signer   the identity keypair authorizing this invocation
+ * @param {string}  networkPassphrase   the network passphrase is incorprated
+ *    into the signature (see {@link Networks} for options)
+ * @param {number}  validUntil  the (exclusive) future ledger sequence number
+ *    until which this authorization entry should be valid (if
+ *    `currentLedgerSeq==validUntil`, this is expired))
+ * @param {xdr.SorobanAuthorizedInvocation} invocation the invocation tree that
+ *    we're authorizing (likely, this comes from transaction simulation)
+ *
+ * @returns {xdr.SorobanAuthorizationEntry}  an authorization entry that you can
+ *    pass along to {@link Operation.invokeHostFunction}
+ */
+export function authorizeInvocation(
+  signer,
+  networkPassphrase,
+  validUntil,
+  invocation
+) {
+  const preimage = buildAuthEnvelope(networkPassphrase, validUntil, invocation);
+  const input = hash(preimage.toXDR());
+  const signature = signer.sign(input);
```

Comment:
We talked about this a bit but I wanted to ask - is there any reason to not support the async case here? I believe if we used an async signature and changed from taking a `signer` to taking a `async(input) => ...sign input and return...` then both sync/async cases would be supported. Something like this -

```
export async function authorizeInvocation(
  signHash,
  networkPassphrase,
  validUntil,
  invocation
) {
  const preimage = buildAuthEnvelope(networkPassphrase, validUntil, invocation);
  const input = hash(preimage.toXDR());
  const signature = await signHash(input);
  ... rest of it ...
```

Reply:
and simple keypair sync signing case can just pass in a `(input) => signer.sign(input)`

Reply:
This method is cleaner because with the callback you need to pass a public key, heh, but I added `authorizeInvocationCallback` for this purpose!

```typescript

export function authorizeInvocationCallback(
  publicKey: string,
  signingMethod: (input: Buffer) => Buffer,
  networkPassphrase: string,
  validUntil: number,
  invocation: xdr.SorobanAuthorizedInvocation
): xdr.SorobanAuthorizationEntry;
```